### DISCUSSION TITLE
Indices of the outer surface cells

### DISCUSSION THREAD

**Discussion by AsimAl-Sofi** [pyvista/pyvista#4387] [Q&A]
Hi all
I would like to know if it is possible in Pyvista to determine the indices of the cells that form the outer surface of a Structured Grid together with its corresponding cell face indices.
for example: 
```py
mesh=out = pv.CylinderStructured(radius=(0,1,2,3),theta_resolution=10,height=5,z_resolution=4)
```
Thanks

**Comment by banesullivan** [comment]
We can use the `extract_surface()` or `extract_geometry()` filters to get the outer surface of a mesh. Specifically, `extract_surface()` will generate new arrays called `vtkOriginalCellIds` and `vtkOriginalPointIds` to track the Cell and Point IDs respectively. You can leverage this like:

https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.extract_surface.html#extract-surface

```py
surface = mesh.extract_surface()
cell_indices = surface['vtkOriginalCellIds']

# Use cell indices to do something to the original data
mesh.cell_data['data'] = np.zeros(mesh.n_cells)
mesh.cell_data['data'][cell_indices] = 1.0
```

Unfortunately, the `CylinderStructured` method's output is a bit unique, and the structured mesh connects back on itself and it isn't a closed loop. So some external faces identified by `extract_surface()` are what we expect to be internal. I tried using a recent addition in PyVista (see https://github.com/pyvista/pyvista/pull/4367) to clean the mesh but it did not work...

We nee to find a way to convert this properly to an `UnstructuredGrid` in order to properly close it

**Reply by akaszynski** [reply]
[clean](https://docs.pyvista.org/version/dev/api/core/_autosummary/pyvista.UnstructuredGridFilters.clean.html#pyvista.UnstructuredGridFilters.clean) indeed works, but you do need to increase the `tolerance`:

```py
import pyvista as pv

out = pv.CylinderStructured(
    radius=(0, 1, 2, 3), theta_resolution=10, height=5, z_resolution=4
)
out.cast_to_unstructured_grid().clean(tolerance=0.001).plot()
```

![sc](https://user-images.githubusercontent.com/11981631/236549103-cc4c9b74-267e-4d1e-8565-372937d77b1e.png)

This feature was just released in `pyvista==0.39.0` available on [PyPI](https://pypi.org/project/pyvista/).

**Reply by banesullivan** [reply]
Ah! 🤦🏻‍♂️  I didn't alter the `tolerance` -- thanks, @akaszynski!

**Reply by akaszynski** [reply]
I'm surprised that `tolerance` is needed at all.

**Reply by banesullivan** [reply]
So a full solution with version 0.39 of PyVista is:

```py
import numpy as np
import pyvista as pv

mesh = pv.CylinderStructured(radius=(0,1,2,3),
                             theta_resolution=10,
                             height=5,
                             z_resolution=4)
# Clean the mesh
mesh = mesh.cast_to_unstructured_grid().clean(tolerance=0.001)

surface = mesh.extract_surface()
cell_indices = surface['vtkOriginalCellIds']
```

```py
# Use cell indeices to do something to original data
mesh.cell_data['data'] = np.zeros(mesh.n_cells)
mesh.cell_data['data'][cell_indices] = 1.0
```

**Reply by AsimAl-Sofi** [reply]
thank you very much guys
I get for some cell indices more than once, does that mean that these cells have 2 or more faces that form the outer surface?

**Reply by banesullivan** [reply]
Correct!

**Reply by AsimAl-Sofi** [reply]
Is it possible to get the corresponding indices of the cell's faces? For example, cell 77 has two faces which are part of the mesh outer face, but what is their indices?

**Reply by AsimAl-Sofi** [reply]
Any suggestions/possible solutions?

**Reply by banesullivan** [reply]
That's a good question and I don't think we have a good way of handling this at this time. We'd need to split the faces of all the cells to be their own cells and track those independently before running the extract surface I think

**Reply by AsimAl-Sofi** [reply]
Thanks
Is there a possibility to implement this functionality in the new Pyvista versions?

**Reply by banesullivan** [reply]
We'd certainly encourage and welcome this new feature! But it will need a champion behind it and concrete use cases
